API: Create ``PyArray_DescrProto`` for legacy descriptor registration

seberg opened the pull request:
First step, force users of registered dtypes to slightly change their code in order to free up changing our `descr` struct.

seberg committed:
API: Create `PyArray_DescrProto` for legacy descriptor registration

ngoldbaum reviewed (commented):
Code changes look fine to me. Ping @WarrenWeckesser too for the perspective of a person maintaining some legacy dtypes.

We don't need to actually write new docs now, but we should either delete https://numpy.org/doc/stable/user/c-info.beyond-basics.html#user-defined-data-types or add a big warning over it saying it's a legacy API.

seberg commented:
Also @jakevdp, `mltypes` will notice this.  The updates should be very easy though (see release notes or `rational` changes here).

I would like to push this through very soon, though.  Without this there is absolutely no way we might make `itemsize` an `intp` or add new flags in a reasonable way, and I think the changes required are simple.
(And however we do it exactly, prepare for bit-sized dtypes, which is asked for surprisingly often, even if `ndarray` can't store them directly due to byte strides.)

seberg commented:
Rebased and added the alias also to the compat header.  I will merge this once tests pass since Nathan said it looks good and it is just a bit nicer to work off main.
